internal/api_server/agentserver/grpc.go

```diff
@@ -0,0 +1,168 @@
+	s.log.Printf("Initializing Agent-side gRPC server: %s", s.cfg.Service.AgentGrpcAddress)
+	tlsCredentials := credentials.NewTLS(s.tlsConfig)
+	server := grpc.NewServer(grpc.Creds(tlsCredentials))
+	pb.RegisterRouterServiceServer(server, s)
+
+	listener, err := net.Listen("tcp", s.cfg.Service.AgentGrpcAddress)
+
+	if err != nil {
+		s.log.Fatalf("cannot start server: %s", err)
+	}
+
+	go func() {
+		<-ctx.Done()
+		server.Stop()
+	}()
+
+	return server.Serve(listener)
+}
+
+type streamCtx struct {
+	cancel context.CancelFunc
+	stream pb.RouterService_StreamServer
+	closed bool // one side closed the connection and we should not accept any more messages
+}
+
+func (s *AgentGrpcServer) Stream(stream pb.RouterService_StreamServer) error {
+	ctx := stream.Context()
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return status.Error(codes.InvalidArgument, "missing metadata")
+	}
+
+	sessionIds := md.Get(SessionIDKey)
+	if len(sessionIds) != 1 {
+		return status.Error(codes.InvalidArgument, "missing "+SessionIDKey)
+	}
+	sessionId := sessionIds[0]
+
+	// this should eventually come from the TLS context
+	clientNames := md.Get(ClientNameKey)
+	if len(clientNames) != 1 {
+		return status.Error(codes.InvalidArgument, "missing "+ClientNameKey)
+	}
+	clientName := clientNames[0]
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	sctx := streamCtx{
+		cancel: cancel,
+		stream: stream,
+		closed: false,
+		// TODO: Add expiration fields so we can clean up old streams ver time
+	}
+
+	actual, loaded := s.pendingStreams.LoadOrStore(sessionId, sctx)
+	// if the map already had a value, we are the second client, so we can start the forwarding
+	// between both clients
+	if loaded {
+		s.log.Infof("client %s connected to session %s", clientName, sessionId)
+		otherSideStream := actual.(streamCtx).stream
+		if actual.(streamCtx).closed || otherSideStream == nil {
+			// the other side closed the connection, we should not accept any more messages
+			s.log.Infof("client %s, attempted connection to %s which was already closed", sessionId)
+			return nil
+		}
+		err := forward(ctx, stream, otherSideStream)
+		if errors.Is(err, io.EOF) {
+			// one side closed the connection, we should not accept any more messages
+			s.log.Infof("one client disconnected from session %s, closing", sessionId)
+
+			// we try to send a close message to both sides, no error checking, best effort
+			stream.Send(&pb.StreamResponse{Closed: true})
+			otherSideStream.Send(&pb.StreamResponse{Closed: true})
+
+			// free the other stream reference, and mark this one as closed
+			stctx := actual.(streamCtx)
+			stctx.stream = nil
+			stctx.closed = true
+			s.pendingStreams.Store(sessionId, stctx)
+			return nil
+		} else {
+			actual.(streamCtx).cancel()
+			return err
+		}
+	} else {
+		// the map did not have a value, we are the first client, so we wait for the second
+		s.log.Infof("client %s waiting for peer %s", clientName, sessionId)
```

Comment:
What happens if the other side never connects? Will the first gRPC connection leak?

Reply:
Yes, we will need to add timeouts in here 👍